nengo/spa/spa.py

```diff
@@ -86,6 +133,20 @@ def __setattr__(self, key, value):
 
             value.on_add(self)
 
+    def __exit__(self, dummy_exc_type, dummy_exc_value, dummy_tb):
+        super(SPA, self).__exit__(dummy_exc_type, dummy_exc_value, dummy_tb)
+        module_list = self._modules.values()
```

Comment:
By changing this to
```python
module_list = set(self._module.values())
```
the code below should become amortized linear. (Maybe it is even better to use `frozenset` instead of `set` because the set won't be changed?)